File: Java/Data Structures & Algorithms Module - Java/Week3/DoublyLinkedList/DoublyLinkedList.java
/*
DoublyLinkedList.java
Billy Cussen
11/09/2020 - Refactored on 12/09, 13/09 & 14/09 as Double Link was not working correctly
*/

public class DoublyLinkedList<T> implements DoublyLinkedListInterface<T>{

    DLLNode<T> first;
    DLLNode<T> last;

    @Override
    public boolean isEmpty(){
    return first==null;
    }

    @Override
    public int size(){
        return sizeHelper(first);
    }

    @Override
    public void addByElement(T element){
        if(isEmpty()){
            first = new DLLNode<T>(element,null,null);
            last = first;
        } else{
            DLLNode<T> current = first;
            while(current.successor!=null){
                current = current.successor;
            }
            if(last == current){
                current.successor = new DLLNode<T>(element,current.successor,null);
                last = current.successor;
                last.previous = current;
            }
        }
    }
    
    @Override
    public void addByIndex(T element, int index){
        if(isEmpty()||index==size()){
            addByElement(element);
        } else if(index < 0 || index > size()){
            throw new NullPointerException();
        } else if(index == 0){
            first = new DLLNode<T>(element,null,first);
            first.successor.previous = first;
        } else{
            DLLNode<T> newNode = new DLLNode<T>(element);
            DLLNode<T> rightNode = getNodeByIndex(index);
            DLLNode<T> leftNode = rightNode.previous;
            leftNode.successor = newNode;
			newNode.previous = leftNode;
			newNode.successor = rightNode;
			rightNode.previous = newNode;
        }
    }

    @Override
    public void addFirst(T element){
        addByIndex(element,0);
    }
    
    @Override
    public void addAtEnd(T element){
        addByIndex(element,size());
    }
    
    @Override
    public void addBefore(T element, T target){
        int index = getIndex(target);
        addByIndex(element,index);
    }
    
    @Override
    public T getByIndex(int index){
        DLLNode<T> current = first;
        for(int i = 0; i < index - 1; i++){
            current = current.successor;
        }
        return current.successor.element;
    }
    
    @Override
    public T removeByElement(T element){
        int index = getIndex(element);
        return removeByIndex(index);
    }
    
    @Override
    public T removeByIndex(int index){
        if(isEmpty()||index<0||index>size()){
            throw new NullPointerException();
        } else if(index == 0){
            T element = first.element;
            first = first.successor;
            first.previous = null;
            return element;
        }else{
            DLLNode<T> current = first;
            for(int i = 0; i < index - 1; i++){
                current = current.successor;
            }
            T element = current.successor.element;
            current.successor = current.successor.successor;
            if (current.successor == null) {
                last = current;
            } else{
                current.successor.previous = current;
            }

            return element;
        }
        
    }

    @Override
    public String toStringBackward(){
        DLLNode<T> current = last;
        String result = new String();
        int position = size()-1;

        return toStringHelper1(current, result, position);
    }

    @Override
    public String toStringForward(){
        DLLNode<T> current = first;
        String result = new String();
        int position = 0;

        return toStringHelper2(current, result, position);
    }

    private int sizeHelper(DLLNode<T> current){
        if(current==null){
            return 0;
        }
        return 1 + sizeHelper(current.successor);
    } 

    private int getIndex(T target){
        DLLNode<T> current = first;
        int index = 0;
        while(current!=null){
            if(current.element.equals(target)){
                return index;
            }
            index++;
            current = current.successor;
        }
        return -1;
    }

    private DLLNode<T> getNodeByIndex(int index){
        DLLNode<T> current = first;
        for(int i = 0; i < index ; i++){
            current = current.successor;
        }
        return current;
    }

    private String toStringHelper1(DLLNode<T> current, String result, int position){
        if(current == null){
            return result;
        }
        result+="Position "+position+", Element: "+current.element+"\n";
        position--;
        return toStringHelper1(current.previous, result, position);
    }

    private String toStringHelper2(DLLNode<T> current, String result, int position){
        if(current == null){
            return result;
        }
        result+="Position "+position+", Element: "+current.element+"\n";
        position++;
        return toStringHelper2(current.successor, result, position);
    }

}

File: Python/Data Structures & Algorithms Module - Python/Week8/DoublyLinkedList.py
"""
DoublyLinkedList.py
Billy Cussen
24/11/2020
"""
from DLNode import DLNode

class DoublyLinkedList:
    def __init__(self):
        self.first = None
        self.last = None

    def addByElement(self,elem):
        if self.isEmpty():
            self.first = DLNode(elem,None,None)
            self.last = self.first
        else:
            curr = self.first
            while curr.succ != None:
                curr = curr.succ
            curr.succ = DLNode(elem,curr,None)
            curr.succ.prev = curr
            self.last = curr.succ
            

    def addByIndex(self, elem, index):
        if index < 0 or index > self.size():
            raise Exception("Can't add Element, Index is Invalid!")
        elif self.isEmpty():
            raise Exception("Can't Add By Index, List is Empty!")
        elif index == self.size():
            self.addByElement(elem)
        elif index == 0:
            self.first = DLNode(elem,None,self.first)
            self.first.succ.prev = self.first
        else:
            curr = self.first
            for i in range(index-1):
                curr = curr.succ
            leftNode = curr
            rightNode = curr.succ
            newNode = DLNode(elem,None,None)
            newNode.prev = leftNode
            leftNode.succ = newNode
            newNode.succ = rightNode
            rightNode.prev = newNode

    def addAfter(self, elem, target):
        index = self.getIndex(target)
        self.addByIndex(elem,index+1)

    def addBefore(self, elem, target):
        index = self.getIndex(target)
        self.addByIndex(elem,index)

    def getIndex(self, target):
        if self.isEmpty():
            raise Exception("Can't get Index, List is Empty!")
        else:
            curr = self.first
            index = 0
            while(curr != None):
                if curr.elem == target:
                    return index
                index+=1
                curr = curr.succ
            return -1


    def size(self):
        if self.isEmpty():
            raise Exception("Can't get Size, DLL is Empty!")
        return self.sizeHelper(self.first)

    def sizeHelper(self,curr):
        if curr == None:
            return 0
        return 1 + self.sizeHelper(curr.succ)

    def isEmpty(self):
        return self.first == None

    def removeByElement(self,target):
        index = self.getIndex(target)
        if index == -1:
            raise Exception("Can't remove element, it does not exist in the List!")
        else:
            return self.removeByIndex(index)

    def removeByIndex(self,index):
        if index >= self.size() or index < 0:
            raise Exception("Can't remove Element, index is Invalid")
        elif self.isEmpty():
            raise Exception("Can't remove Element, List is Empty!")
        elif index == 0:
            elem = self.first.elem
            self.first = self.first.succ
            if self.isEmpty():
                self.last = None
            else:
                self.first.prev = None
            return elem
        else:
            curr = self.first
            for i in range(index-1):
                curr = curr.succ
            elem = curr.succ.elem
            curr.succ = curr.succ.succ
            if curr.succ == None:
                self.last = curr
            else:
                curr.succ.prev = curr
            return elem

    def toString(self):
        if self.isEmpty():
            raise Exception("Can't traverse, DLL is Empty!")
        result = "***DOUBLY LINKED LIST TRAVERSAL***\n"
        return self.toStringHelper(self.first,result,0)

    def toStringHelper(self, curr, result, position):
        if curr == None:
            result+="***END***\n"
            return result
        result+="Position "+str(position)+", Element: "+str(curr.elem)+"\n"
        position+=1
        return self.toStringHelper(curr.succ,result,position)

    def toStringBackwards(self):
        if self.isEmpty():
            raise Exception("Can't do backwards traversal, DLL is Empty!")
        result = "***DOUBLY LINKED LIST BACKWARDS TRAVERSAL***\n"
        return self.toStringBackwardsHelper(self.last,result,self.size()-1)

    def toStringBackwardsHelper(self, curr, result, position):
        if curr == None:
            result+="***END***\n"
            return result
        result+="Position "+str(position)+", Element: "+str(curr.elem)+"\n"
        position-=1
        return self.toStringBackwardsHelper(curr.prev,result,position)